### PR TITLE
Repopulate a StepByStepPage with content from publishing-api

### PR DESCRIPTION
**Pull request by leenagupte**
Trello: https://trello.com/c/c4z7hAxs

Dependant on PR #487

## Motivation

When draft changes are made to a published step by step in collections publisher, there is no way to discard those draft changes without manually unpicking them.

However, all of the previous editions are being stored in Publishing API.

We should give publishers the ability to discard a draft in collections-publisher by getting the latest published version of the step by step from the publishing-api and using that to repopulate the data for the step by step in collections-publisher.

In other words, programmatically unpicking the changes so that the publisher doesn't have to.

## What's changed

This is the first step in the ability to overwrite step by step data in collections-publisher with a version of the step by step held by publishing-api. The service takes any step by step payload from publishing-api and uses it to update the step by step in the database.

### CODE REVIEW

File: app/services/step_by_step_page_reverter.rb [hunk 1]
```diff
@@ -52,4 +53,12 @@ def steps
   def logic(step)
     step[:logic] || "number"
   end
+
+  def contents(step_contents)
+    contents_list = step_contents.map do |content|
+      content[:text]
+    end
+
+    contents_list.join("\r\n")
```
Comment (sihugh) [inline app/services/step_by_step_page_reverter.rb]: I'm interested in this new line.  I guess it works?  Does `\n` work too?

File: spec/services/step_by_step_page_reverter_spec.rb [hunk 2]
```diff
@@ -83,6 +83,12 @@
         expect(step_by_step_page.steps[1].position).to eq(2)
         expect(step_by_step_page.steps[2].position).to eq(3)
       end
+
+      describe "#contents" do
+        it "saves the contents of a paragraph" do
+          expect(step_by_step_page.steps[0].contents).to eq("A paragraph of text in the first step.")
```
Comment (sihugh) [inline spec/services/step_by_step_page_reverter_spec.rb]: Can we test for the multiline thing?

File: app/models/step.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@ class Step < ApplicationRecord
   belongs_to :step_by_step_page
   validates_presence_of :step_by_step_page
   validates :title, :logic, presence: true
-  has_many :link_reports
+  has_many :link_reports, :dependent => :destroy
```
Comment (sihugh) [inline app/models/step.rb]: Same comment as on the other PR - should we be explicit about this in a model test?

File: app/services/step_by_step_page_reverter.rb [hunk 4]
```diff
@@ -11,6 +11,7 @@ def repopulate_from_publishing_api
     step_by_step_page.slug = payload_from_publishing_api[:base_path].tr('/', '')
     step_by_step_page.introduction = introduction
     step_by_step_page.description = payload_from_publishing_api[:description]
+    step_by_step_page.updated_at = public_updated_at
```
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/services/step_by_step_page_reverter.rb]: Hmmm. Perhaps not. I'll take it out.

File: app/services/step_by_step_page_reverter.rb [hunk 5]
```diff
@@ -12,6 +12,7 @@ def repopulate_from_publishing_api
     step_by_step_page.introduction = introduction
     step_by_step_page.description = payload_from_publishing_api[:description]
     step_by_step_page.updated_at = public_updated_at
+    step_by_step_page.published_at = public_updated_at
```
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/services/step_by_step_page_reverter.rb]: I think you're right. I was being overly cautious. I'll take it out. Cheers!

File: app/services/step_by_step_page_reverter.rb (hunk 1 above)
Comment (leenagupte) [reply app/services/step_by_step_page_reverter.rb]: @sihugh I looked at how steps are stored in the database and just copied that!

File: app/models/step.rb (hunk 3 above)
Comment (leenagupte) [reply app/models/step.rb]: I cherry-picked in the commit from the other PR to get the tests to work. I'll unpick it when the other PR is done/merged.

File: app/services/step_by_step_page_reverter.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/services/step_by_step_page_reverter.rb]: I'll change the test to match the one for `created_at`.

File: app/services/step_by_step_page_reverter.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/services/step_by_step_page_reverter.rb]: I'll add a test to check it isn't overwritten ala `created_at`

File: spec/services/step_by_step_page_reverter_spec.rb (hunk 2 above)
Comment (leenagupte) [reply spec/services/step_by_step_page_reverter_spec.rb]: I have a test for a multi-line, multi-list, multi-bullets, with and without context monstrosity of a step in a later commit.

File: app/services/step_by_step_page_reverter.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/services/step_by_step_page_reverter.rb]: Actually, I think you're @sihugh and it's not needed at all. `updated_at` is updated every time the step nav is saved, so the value will have changed again when the publishers are ready to save and push to publishing-api.

File: app/services/step_by_step_page_reverter.rb (hunk 1 above)
Comment (leenagupte) [reply app/services/step_by_step_page_reverter.rb]: Actually @sihugh, I think I need to join on a double newline, so that paragraphs and lists have an empty line between them.

**Comment by leenagupte** [comment]
@sihugh I've addressed your comments and fixed up. Github is showing the commits in an odd order though.

They should be ordered thusly:
```
Repopulate title of step nav from publishing-api
Repopulate the slug of the step by step
Repopulate the introduction of the step by step
Repopulate the description of a step by step
Don’t repopulate the content-id for a step by step
Don’t repopulate the created date of the step by step
Don't repopulate the published_at date of the step by step
Repopulate draft_updated_at time with public_updated_at
Populate the logic for each step
Populate if the steps in the step by step are optional
Populate the position of each step
Saves the contents of a step that is just text
Saves the contents of a step that contains links
Saves contents of step with bulleted list of links
Saves contents of step with bulleted list of text
Saves lists of links with context
Saves bulleted lists of links with context
Check content can save links and paragraphs
Save Navigation Rules for all internal links
Turn off step nav for rules in `pages_related_to_step_nav`
```

File: spec/services/step_by_step_page_reverter_spec.rb (hunk 2 above)
Comment (leenagupte) [reply spec/services/step_by_step_page_reverter_spec.rb]: @sihugh I've updated the test to store multiple paragraphs. I felt it was appropriate as I mentioned it in the comments. Cheers!

**Comment by leenagupte** [comment]
@sihugh It's the next story!

**Comment by sihugh** [comment]
MAYBE JUST RUN IT AS A RAKE TASK AT RANDOM INTERVALS 👿

**Comment by leenagupte** [comment]
@sihugh :trollface:
